File: Source/Events/WithinRange/UnitWithinRange.cs
using System;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class UnitWithinRange
{
    private static bool RegisterUnitWithinRangeSuper(unit u, float range, bool cleanOnKilled, Func<bool> filter, trigger execution)
    {
        if (range <= 0)
        {
            return false;
        }

        TriggerRegisterUnitInRange(execution, u, range, Condition(filter));

        return true;
    }

    public static void DeRegisterUnitWithinRangeUnit(Kitty kitty)
    {
        kitty.w_Collision.ClearActions();
        kitty.c_Collision.ClearActions();
        kitty.w_Collision.Dispose();
        kitty.c_Collision.Dispose();
        kitty.w_Collision = null;
        kitty.c_Collision = null;
    }

    public static void DeRegisterUnitWithinRangeUnit(ShadowKitty kitty)
    {
        kitty.cCollision.ClearActions();
        kitty.wCollision.ClearActions();
        kitty.wCollision.Dispose();
        kitty.cCollision.Dispose();
        kitty.wCollision = null;
        kitty.cCollision = null;
    }

    public static bool RegisterUnitWithinRangeTrigger(unit u, float range, Func<bool> filter, trigger execution)
    {
        return RegisterUnitWithinRangeSuper(u, range, false, filter, execution);
    }
}

File: src/Events/WithinRange/UnitWithinRange.ts
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { ShadowKitty } from 'src/Game/Entities/ShadowKitty'
import { Trigger, Unit } from 'w3ts'

export class UnitWithinRange {
    private static RegisterUnitWithinRangeSuper = (
        u: Unit,
        range: number,
        cleanOnKilled: boolean,
        filter: () => boolean,
        execution: Trigger
    ) => {
        if (range <= 0) {
            return false
        }

        execution.registerUnitInRage(u.handle, range, Condition(filter))

        return true
    }

    public static DeRegisterUnitWithinRangeUnit = (kitty: Kitty) => {
        if (!kitty.c_Collision || !kitty.w_Collision) return
        kitty.w_Collision.removeActions()
        kitty.c_Collision.removeActions()
        kitty.w_Collision.destroy()
        kitty.c_Collision.destroy()
        kitty.w_Collision = null as never
        kitty.c_Collision = null as never
    }

    public static DeRegisterUnitWithinRangeUnitShadow = (shadowKitty: ShadowKitty) => {
        if (!shadowKitty.cCollision || !shadowKitty.wCollision) return
        shadowKitty.cCollision.removeActions()
        shadowKitty.wCollision.removeActions()
        shadowKitty.wCollision.destroy()
        shadowKitty.cCollision.destroy()
        shadowKitty.wCollision = null as never
        shadowKitty.cCollision = null as never
    }

    public static RegisterUnitWithinRangeTrigger = (
        u: Unit,
        range: number,
        filter: () => boolean,
        execution: Trigger
    ) => {
        return UnitWithinRange.RegisterUnitWithinRangeSuper(u, range, false, filter, execution)
    }
}
